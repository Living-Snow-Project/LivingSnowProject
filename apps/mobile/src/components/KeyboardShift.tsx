import React, { Component } from "react";
import {
  Animated,
  Dimensions,
  EmitterSubscription,
  Keyboard,
  KeyboardEventListener,
  Platform,
  StyleSheet,
  TextInput,
} from "react-native";

const styles = StyleSheet.create({
  container: {
    height: "100%",
    left: 0,
    position: "absolute",
    top: 0,
    width: "100%",
  },
});

interface IProps {
  children: () => React.ReactNode;
}

interface IState {
  shift: Animated.Value;
}

const { State: TextInputState } = TextInput;

export class KeyboardShift extends Component<IProps, IState> {
  isHiding: Animated.CompositeAnimation | null;

  previousGap: number;

  keyboardDidShowSub: EmitterSubscription;

  keyboardDidHideSub: EmitterSubscription;

  constructor(props) {
    super(props);
    this.state = {
      shift: new Animated.Value(0),
    };
    this.previousGap = 0;
  }

  componentDidMount() {
    if (Platform.OS === "ios") {
      this.keyboardDidShowSub = Keyboard.addListener(
        "keyboardDidShow",
        this.handleKeyboardDidShow
      );
      this.keyboardDidHideSub = Keyboard.addListener(
        "keyboardDidHide",
        this.handleKeyboardDidHide
      );
    }
  }

  componentWillUnmount() {
    if (Platform.OS === "ios") {
      this.keyboardDidShowSub.remove();
      this.keyboardDidHideSub.remove();
    }
  }

  handleKeyboardDidShow: KeyboardEventListener = (event) => {
    const { shift } = this.state;
    const { height: windowHeight } = Dimensions.get("window");
    // when the multiline TextInput grows, we want the keyboard to move with it
    const keyboardHeight = event.endCoordinates.height;
    const currentlyFocusedInput = TextInputState.currentlyFocusedInput();

    if (this.isHiding) {
      this.isHiding.stop();
      this.isHiding = null;
    }

    if (currentlyFocusedInput != null) {
      currentlyFocusedInput.measure(
        (originX, originY, width, height, pageX, pageY) => {
          const fieldHeight = height;
          const fieldTop = pageY;
          let gap =
            windowHeight - keyboardHeight - (fieldTop + fieldHeight + 10);

          if (!gap) {
            return;
          }

          // negative gap means the currentlyFocusedInput is covered by Keyboard
          if (gap < 0) {
            gap += this.previousGap;
          }

          // positive gap means the currentlyFocusedInput is not covered by Keyboard
          if (gap >= 0) {
            return;
          }

          this.previousGap = gap;
          Animated.timing(shift, {
            toValue: gap,
            duration: 250,
            useNativeDriver: true,
          }).start();
        }
      );
    }
  };

  handleKeyboardDidHide = () => {
    const { shift } = this.state;
    this.isHiding = Animated.timing(shift, {
      toValue: 0,
      duration: 250,
      useNativeDriver: true,
    });

    // it is common, and unpredictable, that 'keyboardDidHide' events are fired even though the keyboard remained visible
    // if this happens and 'keyboardDidShow' is simultaneously fired, we stop hiding animation from completing
    this.isHiding.start(({ finished }) => {
      this.isHiding = null;
      if (finished) {
        this.previousGap = 0;
      }
    });
  };

  render() {
    const { children } = this.props;
    const { shift } = this.state;

    return (
      <Animated.View
        style={[styles.container, { transform: [{ translateY: shift }] }]}
      >
        {children()}
      </Animated.View>
    );
  }
}
