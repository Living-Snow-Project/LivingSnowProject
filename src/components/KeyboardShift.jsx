import { PropTypes } from "prop-types";
import React, { Component } from "react";
import {
  Animated,
  Dimensions,
  Keyboard,
  Platform,
  StyleSheet,
  TextInput,
} from "react-native";

const { State: TextInputState } = TextInput;

export default class KeyboardShift extends Component {
  constructor(props) {
    super(props);
    this.state = {
      shift: new Animated.Value(0),
    };
    this.previousGap = 0;
  }

  componentDidMount() {
    if (Platform.OS === "ios") {
      this.keyboardDidShowSub = Keyboard.addListener(
        "keyboardDidShow",
        this.handleKeyboardDidShow
      );
      this.keyboardDidHideSub = Keyboard.addListener(
        "keyboardDidHide",
        this.handleKeyboardDidHide
      );
    }
  }

  componentWillUnmount() {
    if (Platform.OS === "ios") {
      this.keyboardDidShowSub.remove();
      this.keyboardDidHideSub.remove();
    }
  }

  handleKeyboardDidShow = (event) => {
    const { shift } = this.state;
    const { height: windowHeight } = Dimensions.get("window");
    // when the multiline TextInput grows, we want the keyboard to move with it
    const keyboardHeight =
      event?.endCoordinates?.height === undefined
        ? this.keyboardHeight
        : event.endCoordinates.height;
    this.keyboardHeight = keyboardHeight;
    const currentlyFocusedInput = TextInputState.currentlyFocusedInput();

    if (this.isHiding) {
      this.isHiding.stop();
      this.isHiding = null;
    }

    if (currentlyFocusedInput != null) {
      currentlyFocusedInput.measure(
        (originX, originY, width, height, pageX, pageY) => {
          const fieldHeight = height;
          const fieldTop = pageY;
          let gap = windowHeight - keyboardHeight - (fieldTop + fieldHeight);

          if (!gap) {
            return;
          }

          // negative gap means the currentlyFocusedInput is covered by Keyboard
          if (gap < 0) {
            gap += this.previousGap;
          }

          // positive gap means the currentlyFocusedInput is not covered by Keyboard
          if (gap >= 0) {
            return;
          }

          this.previousGap = gap;
          Animated.timing(shift, {
            toValue: gap,
            duration: 250,
            useNativeDriver: true,
          }).start();
        }
      );
    }
  };

  handleKeyboardDidHide = () => {
    const { shift } = this.state;
    this.isHiding = Animated.timing(shift, {
      toValue: 0,
      duration: 250,
      useNativeDriver: true,
    });

    // it is common, and unpredictable, that 'keyboardDidHide' events are fired even though the keyboard remained visible
    // if this happens and 'keyboardDidShow' is simultaneously fired, we stop hiding animation from completing
    this.isHiding.start(({ finished }) => {
      this.isHiding = null;
      if (finished) {
        this.previousGap = 0;
      }
    });
  };

  render() {
    const { children: renderProp } = this.props;
    const { shift } = this.state;

    return (
      <Animated.View
        style={[styles.container, { transform: [{ translateY: shift }] }]}
      >
        {renderProp()}
      </Animated.View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    height: "100%",
    left: 0,
    position: "absolute",
    top: 0,
    width: "100%",
  },
});

KeyboardShift.propTypes = {
  children: PropTypes.func.isRequired,
};
