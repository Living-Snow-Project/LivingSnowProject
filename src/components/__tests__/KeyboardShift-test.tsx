import React from "react";
import { Keyboard, Platform, TextInput } from "react-native";
import { render } from "@testing-library/react-native";
import KeyboardShift from "../KeyboardShift";

const { State: TextInputState } = TextInput;

describe("KeyboardShift test suite", () => {
  test("renders on ios", () => {
    const { toJSON } = render(
      <KeyboardShift>{() => <TextInput />}</KeyboardShift>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  test("renders on android", () => {
    Platform.OS = "android";
    const { toJSON, unmount } = render(
      <KeyboardShift>{() => <TextInput />}</KeyboardShift>
    );

    expect(toJSON()).toMatchSnapshot();

    unmount();

    Platform.OS = "ios";
  });

  test("emit keyboard events", async () => {
    const event = {
      endCoordinates: {
        height: 50,
      },
    };

    render(<KeyboardShift>{() => <TextInput />}</KeyboardShift>);

    // the best we can do is execute the code and make sure it doesn't crash
    Keyboard.emit("keyboardDidShow");

    jest
      .spyOn(TextInputState, "currentlyFocusedInput")
      // @ts-ignore
      .mockImplementation(() => ({
        measure: (cb) => {
          cb(0, 0, 0, 50, 0, 50);
        },
      }));

    // gap = NaN
    Keyboard.emit("keyboardDidShow");

    // gap >= 0
    Keyboard.emit("keyboardDidShow", event);
    Keyboard.emit("keyboardDidHide");

    event.endCoordinates.height = 5000;
    // gap < 0
    Keyboard.emit("keyboardDidShow", event);
    Keyboard.emit("keyboardDidHide");

    // wait for the keyboard hiding animation to finish
    await new Promise((r) => {
      setTimeout(r, 250);
    });
  });
});
