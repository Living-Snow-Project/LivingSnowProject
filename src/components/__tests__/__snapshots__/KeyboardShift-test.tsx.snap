// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyboardShift test suite renders on android 1`] = `
<View
  style={
    Object {
      "height": "100%",
      "left": 0,
      "position": "absolute",
      "top": 0,
      "transform": Array [
        Object {
          "translateY": 0,
        },
      ],
      "width": "100%",
    }
  }
>
  <TextInput
    allowFontScaling={true}
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
  />
</View>
`;

exports[`KeyboardShift test suite renders on ios 1`] = `
<View
  style={
    Object {
      "height": "100%",
      "left": 0,
      "position": "absolute",
      "top": 0,
      "transform": Array [
        Object {
          "translateY": 0,
        },
      ],
      "width": "100%",
    }
  }
>
  <TextInput
    allowFontScaling={true}
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
  />
</View>
`;
